File: samples-java/src/main/java/io/github/ilaborie/loops/samples/recursion/FilterTR.java
package io.github.ilaborie.loops.samples.recursion;

import io.github.ilaborie.loops.samples.Element;

import java.util.ArrayList;
import java.util.List;

import static io.github.ilaborie.loops.samples.Elements.getSomeElements;

public class FilterTR {

    public static void main(String[] args) {
        List<Element> elements = getSomeElements();
        List<Element> result = filter(elements);
        result.forEach(System.out::println);
    }

    public static List<Element> filter(List<Element> elements) {
        return aux(elements, new ArrayList<>());
    }

    public static List<Element> aux(List<Element> elements, List<Element> acc) {
        if (elements.isEmpty()) {
            return acc;
        }

        Element head = elements.get(0);
        List<Element> tail = elements.subList(1, elements.size());

        if (head.getName().startsWith("a")) {
            acc.add(head);
        }
        return aux(tail, acc);
    }

}

File: samples-kotlin/src/main/kotlin/io/github/ilaborie/loops/samples/recursion/filterTR.kt
package io.github.ilaborie.loops.samples.recursion

import io.github.ilaborie.loops.samples.ElementKt
import io.github.ilaborie.loops.samples.getSomeElementsKt


fun main() {
    val elements = getSomeElementsKt()
    val result = filterTr(elements)
    result.forEach { println(it) }
}

fun filterTr(elements: List<ElementKt>): List<ElementKt> {
    tailrec fun aux(elements: List<ElementKt>, acc: List<ElementKt>): List<ElementKt> {
        if (elements.isEmpty()) {
            return acc
        }

        val head = elements.first()
        val tail = elements.drop(1)

        return if (head.name.startsWith("a")) {
            aux(tail, acc + head)
        } else {
            aux(tail, acc)
        }
    }

    return aux(elements, emptyList())
}

fun filterTrAlt(elements: List<ElementKt>): List<ElementKt> {
    tailrec fun aux(elements: List<ElementKt>, acc: List<ElementKt>): List<ElementKt> =
        when {
            elements.isEmpty()                    ->
                acc
            elements.first().name.startsWith("a") ->
                aux(elements.drop(1), acc + elements.first())
            else                                  ->
                aux(elements.drop(1), acc)
        }

    return aux(elements, emptyList())
}
